### ISSUE TITLE
AVS authentication issue

### ISSUE THREAD

**Issue by clouduser22** [Azure/azure-sdk-for-java#26525]
**Query/Question**
While trying to authenticate into AVS with the service principal credentials, I get this error:
"exceptionInfo":"com.azure.core.management.exception.ManagementException: Status code 403, \\"{\\"error\\":{\\"code\\":\\"AuthorizationFailed\\",\\"message\\":\\"
The client <clientId> with object id <clientId> does not have authorization to perform action 'Microsoft.AVS/privateClouds/read' over scope '/subscriptions/<subscriptionId>' or the scope is invalid. If access was recently granted, please refresh your credentials.

I am using ClientSecretCredentialBuilder() class and passing the tenantId, clientId and clientSecret from the service principal.

***Why is this not a Bug or a feature Request?***
It is likely some step or privilege I am missing. I am looking for pointers on what I might be missing.

**Setup (please complete the following information if applicable):**
 - OS: Linux, custom Azure machine image
 - IDE: [e.g. IntelliJ]
 - Library/Libraries: com.azure:azure-identity:1.4.1
 
 **Information Checklist**
 Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [ ] Query Added
- [ ] Setup information Added

**Comment by joshfree** [comment]
@billwert could you please assist @clouduser22

**Comment by clouduser22** [comment]
@weidongxu-microsoft Thanks for looking into this.

I am using the AzureProfile constructor that takes both tenantId and subscription id:
AzureProfile profile = new AzureProfile(tenantId, subscriptionId, AzureEnvironment.AZURE);

I am not able to login with the service principal with az login:
# az login --service-principal -t <tenantId> -u <appId> -p <client secret>
No subscriptions found for <appId>

I did create the service principal by following (https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal). But, for some reason, it says a subscription is not found for this appId. Do you know why it might happen? Is it the right way to create the principal to access AVS?

**Comment by clouduser22** [comment]
@weidongxu-microsoft 
I then tried to use the managed identities as I am authenticating into the AVS from an Azure VM. I enabled the system managed identity on the Azure VM and also did the role assignment to give permissions for 'Microsoft.AVS/privateClouds/read'.

With that, I am able to login through CLI:

# az login --identity 
[
  {
    "environmentName": "AzureCloud",
    "homeTenantId": "xxxx",
    "id": "xxx",
    "isDefault": true,
    "managedByTenants": [],
    "name": "AVS2.0Interop",
    "state": "Enabled",
    "tenantId": "xxxxx",
    "user": {
      "assignedIdentityInfo": "MSI",
      "name": "systemAssignedIdentity",
      "type": "servicePrincipal"
    }
  }
]

and can also read the AVS through CLI:
# az vmware private-cloud list
[
  {
    "availability": {
      "secondaryZone": null,
      "strategy": "SingleZone",
      "zone": null
    },
   ....
}

However, doing the same from Java fails:
AzureProfile profile = new AzureProfile(tenantId, subscriptionId, AzureEnvironment.AZURE);
TokenCredential credential = new ManagedIdentityCredentialBuilder().build()
avsManager = AvsManager.authenticate(credential, profile);

2022-01-19T06:08:14.552+0000 [.qtp984148985-31] MSI: Failed to acquire tokens after retrying 3 times
2022-01-19T06:08:14.556+0000 [.qtp984148985-31] Azure Identity => ERROR in getToken() call for scopes [https://management.core.windows.net//.default]: MSI: Failed to acquire tokens after retrying 3 times
2022-01-19T06:08:14.556+0000 [.qtp984148985-31] Failed to acquire a new access token.

Any ideas on why this might be happening?

**Comment by weidongxu-microsoft** [comment]
@clouduser22 

If both SDK and CLI cannot use the service principal, it is likely problem with RBAC. Your doc is good (and the RBAC for the service principal is almost same to the RBAC for the managed identity).

It is recommended to use managed identity, if the app works in VM. And `ManagedIdentityCredentialBuilder().build()` should work out-of-the-box for the system assigned managed identity.

I probably need to open a VM to try it myself.

Just in case, let me know what role you assigned to the managed identity (Contributor), so I can try reproduce. Though "Failed to acquire a new access token." happens before the check on RBAC.

**Comment by clouduser22** [comment]
@weidongxu-microsoft
ok, I will look more into the service principal based auth later. It would be great if I can get the managed identity based auth work for now. Yes, the client is running on the Azure VM in this case.

I assigned the 'Contributor' role to the (system) managed identity on the subscription. And yes, getting the token itself seems to fail in that case but goes thru' for the CLI somehow. Of course, I am invoking the CLI and the Java code from the same azure instance.

**Comment by weidongxu-microsoft** [comment]
@clouduser22 

Got it. I will try to run this https://github.com/weidongxu-microsoft/az-java-sdk-aad/blob/main/src/main/java/com/microsoft/azure/aad/Main.java in my VM with MSI as Contributor to test.

**Comment by weidongxu-microsoft** [comment]
@clouduser22 

Could you try azure-identity 1.4.3?

I got same problem with 1.4.1, but OK with 1.4.3.
```
[com.microsoft.azure.aad.Main.main()] INFO com.azure.core.implementation.jackson.JacksonVersion - Package versions: jackson-annotations=2.13.1, jackson-core=
2.13.1, jackson-databind=2.13.1, jackson-dataformat-xml=2.13.1, jackson-datatype-jsr310=2.13.1, azure-core=1.24.1, Troubleshooting version conflicts: https:/
/aka.ms/azsdk/java/dependency/troubleshoot
[com.microsoft.azure.aad.Main.main()] INFO com.azure.identity.ManagedIdentityCredential - Azure Identity => Managed Identity environment: AZURE VM IMDS ENDPO
INT
[com.microsoft.azure.aad.Main.main()] INFO com.azure.identity.ManagedIdentityCredential - Azure Identity => getToken() result for scopes [https://management.
core.windows.net//.default]: SUCCESS
[com.microsoft.azure.aad.Main.main()] INFO com.azure.core.implementation.AccessTokenCache - Acquired a new access token.
[com.microsoft.azure.aad.Main.main()] INFO com.microsoft.azure.aad.Main - Number of privateClouds: 0
```

I am at loss here what's wrong with 1.4.1 @g2vinay

**Comment by clouduser22** [comment]
Thanks @weidongxu-microsoft for the triage! I tried with 1.4.3 and it works for me as well.

This will help me to make forward progress, so please feel free to close this issue :-). But, I do have two questions:

1. Is it possible to come up with a working example for service principal based credential auth as well (similar to the one you came up with for MSI)? It would be good to understand how that works.
2. If the app runs inside non-azure VM or resource, what is the recommended auth scheme? Obviously, CLI credentials will work, but that is not an option for the daemon app this is. Is there any other option?

**Comment by weidongxu-microsoft** [comment]
@clouduser22

You can check doc here https://docs.microsoft.com/azure/developer/java/sdk/identity

The recommendation would be the `DefaultAzureCredentialBuilder().build()`. See here https://github.com/Azure/azure-sdk-for-java/tree/main/sdk/identity/azure-identity#defaultazurecredential

Basically it is a chain, e.g. first check ENV, then MSI, then CLI, then..., and pick first one works. So it will get to MSI on your VM, and ENV or CLI on your local test environment.

For service principal, it is included in the `Environment` stage of the `DefaultAzureCredential`. We usually define following env variables for it to pick up.
```
AZURE_TENANT_ID
AZURE_CLIENT_ID
AZURE_CLIENT_SECRET
```
and for mgmt one additional
```
AZURE_SUBSCRIPTION_ID
```

We use this literally everyday. So my guess is that the problem is not how you use the SDK, but the service principal or the RBAC.

**Comment by clouduser22** [comment]
Thanks @weidongxu-microsoft . I will look some more on the app service principal.

**Comment by g2vinay** [comment]
> @clouduser22
> 
> Could you try azure-identity 1.4.3?
> 
> I got same problem with 1.4.1, but OK with 1.4.3.
> 
> ```
> [com.microsoft.azure.aad.Main.main()] INFO com.azure.core.implementation.jackson.JacksonVersion - Package versions: jackson-annotations=2.13.1, jackson-core=
> 2.13.1, jackson-databind=2.13.1, jackson-dataformat-xml=2.13.1, jackson-datatype-jsr310=2.13.1, azure-core=1.24.1, Troubleshooting version conflicts: https:/
> /aka.ms/azsdk/java/dependency/troubleshoot
> [com.microsoft.azure.aad.Main.main()] INFO com.azure.identity.ManagedIdentityCredential - Azure Identity => Managed Identity environment: AZURE VM IMDS ENDPO
> INT
> [com.microsoft.azure.aad.Main.main()] INFO com.azure.identity.ManagedIdentityCredential - Azure Identity => getToken() result for scopes [https://management.
> core.windows.net//.default]: SUCCESS
> [com.microsoft.azure.aad.Main.main()] INFO com.azure.core.implementation.AccessTokenCache - Acquired a new access token.
> [com.microsoft.azure.aad.Main.main()] INFO com.microsoft.azure.aad.Main - Number of privateClouds: 0
> ```
> 
> I am at loss here what's wrong with 1.4.1 @g2vinay

There was an edge case that was fixed with azure-identity 1.4.2 release, so that resolved the issue here.
